Parametric google-test for mapping validation

JoergAtGithub opened the pull request:
For debugging https://github.com/mixxxdj/mixxx/pull/11625#issuecomment-2131243612 I modified the googletest setup for mapping validation to use a parametrized list of mappings.
While the results look nice in my IDE, where I can easily debug a single mapping validation test now:
![grafik](https://github.com/mixxxdj/mixxx/assets/64457745/c910d9c3-3b60-4b1a-9487-de49faa8baf2)

I was disappointed to see that CTest combines all the single GTest instances:

```
        Start 121: */LegacyControllerMappingValidationTest.ValidateMappingXML/*
121/840 Test #121: */LegacyControllerMappingValidationTest.ValidateMappingXML/* ...............................   Passed    0.37 sec
        Start 122: */LegacyControllerMappingValidationTest.LoadMapping/*
122/840 Test #122: */LegacyControllerMappingValidationTest.LoadMapping/* ......................................***Failed    5.44 sec
```

I could solve that by using gtest_discover_tests in the CMakelist.txt, now the CTest output looks like this:
```
          Start 1039: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_CDJ_350_CH2#GetParam()=<MappingInfo>
1039/1107 Test #1039: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_CDJ_350_CH2#GetParam()=<MappingInfo> .................................   Passed    0.45 sec
          Start 1040: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_CDJ_850#GetParam()=<MappingInfo>
1040/1107 Test #1040: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_CDJ_850#GetParam()=<MappingInfo> .....................................   Passed    0.45 sec
          Start 1041: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_DDJ_200#GetParam()=<MappingInfo>
1041/1107 Test #1041: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_DDJ_200#GetParam()=<MappingInfo> .....................................   Passed    0.53 sec
          Start 1042: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_DDJ_400#GetParam()=<MappingInfo>
1042/1107 Test #1042: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_DDJ_400#GetParam()=<MappingInfo> .....................................   Passed    0.45 sec
          Start 1043: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_DDJ_FLX4#GetParam()=<MappingInfo>
1043/1107 Test #1043: MidiMappingsValid/LegacyControllerMappingValidationTest.LoadMapping/Pioneer_DDJ_FLX4#GetParam()=<MappingInfo> ....................................   Passed    0.47 sec
```

JoergAtGithub committed:
Fixed wrong and duplicated mapping name

JoergAtGithub committed:
Fixed wrong and duplicated mapping name

JoergAtGithub committed:
Use parametric GTest for mapping validation, instead of one test for all mappings

JoergAtGithub committed:
Use parametric GTest for mapping validation, instead of one test for all mappings

JoergAtGithub committed:
Improved output by suppressing the hex dump of MappingInfo as parameter
Fixed clazy error

JoergAtGithub committed:
Improved output by suppressing the hex dump of MappingInfo as parameter
Fixed clazy error

JoergAtGithub committed:
Use gtest_discover_tests instead of gtest_add to resolve the parametrized gtests

JoergAtGithub committed:
Use gtest_discover_tests instead of gtest_add to resolve the parametrized gtests

JoergAtGithub committed:
Fixed QT_QPA_PLATFORM setting, which was not defined in build-check matrix

JoergAtGithub committed:
Fixed QT_QPA_PLATFORM setting, which was not defined in build-check matrix

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into parametricGTestForMappings

Swiftb0y commented on src/test/controller_mapping_validation_test.h:
```diff
@@ -136,12 +136,40 @@ class FakeController : public Controller {
+class LegacyControllerMappingList {
+  public:
+    LegacyControllerMappingList() {
+        QList<QString> possiblePaths = {
+                QDir::currentPath() + "/res/controllers",
+                QDir::currentPath() + "/../res/controllers",
+                QDir::currentPath() + "/../../res/controllers"};
```
don't we have `getResourcePath` for that instead of having to guess?

Swiftb0y commented on src/test/controller_mapping_validation_test.cpp:
```diff
@@ -163,35 +159,69 @@ bool lintMappingInfo(const MappingInfo& mapping) {
-        EXPECT_TRUE(lintMappingInfo(mapping)) << errorDescription;
-        EXPECT_TRUE(testLoadMapping(mapping)) << errorDescription;
+// Print the mapping name instead with underscores instead of spaces and special characters
+class PrintMappingName {
+  public:
+    template<class ParamType>
+    std::string operator()(const ::testing::TestParamInfo<ParamType>& info) const {
+        // ParamType is MappingInfo
+        QString nameCopy = info.param.getName();
+        static const QRegularExpression regex("[^\\w\\s]");
+        nameCopy.replace(regex, "_");
+        std::string testName = nameCopy.replace(" ", "_").toStdString();
+        return testName;
     }
+};
```
this seems overcomplicated. why not rely on `PrintToStringParamName`?

Swiftb0y commented on src/test/controller_mapping_validation_test.cpp:
```diff
@@ -163,35 +159,69 @@ bool lintMappingInfo(const MappingInfo& mapping) {
     }
+};
+
+auto pEnumerator = createMappingEnumerator();
```
this relies on running before main. I don't know how safe that is considering the heavy Qt classes involved. It would be better if this was run in `main()`, but before the gtest init code. Is that feasible?

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into parametricGTestForMappings

JoergAtGithub committed:
Removed code that tdepends on mixxxdb mixxxtest. Builds again but runs slow

acolombier committed:
Improve naming for parametrised mapping tests

JoergAtGithub commented:
The problem is that this PR slows down the test significiantly, because creating the list of mappings takes a lot of time. A hardcoded list shouldn't have this issue.
Otherwise thi PR works.

JoergAtGithub commented:
The whole testsuite needs about 2x time as before, because the list of mappings is calculate before every test (not only the test that depend on this parameter)

acolombier committed:
WIP: compute mapping list in CMake

acolombier commented:
I have [pushed a version](https://github.com/mixxxdj/mixxx/commit/13611c3b06#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a) with the mapping list computed directly in CMake. Only did MIDI as example, but keen to hear if we would want to go ahead with it. There is undoubtedly an overhead with running multiple tests, perhaps we could make it optional, so only the `coverage` job would run the test in parametrised mode and the slowdown wouldn't impact the build job?

acolombier commented:
Yes, this is what my solution does. It only finds the mapping at compile time to build the parameter list. Once built, `mixxx-test` can be used independently to CTest, and should work fine in the IDE as it will find the `res/controller` directory in the same way it predictably find `src`

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.6' into parametricGTestForMappings

JoergAtGithub committed:
Set RESOURCE_FOLDER path as compile definition for mixxx-test in CMake
Use CMake to collect mapping files at configure time and pass them compile definition.

JoergAtGithub committed:
Re-add accidentally deleted setup code for QML

JoergAtGithub committed:
Fix clazy

JoergAtGithub committed:
Fix Uncaught exception: file:///home/runner/work/mixxx/mixxx/res/controllers/midi-components-0.0.js:140:\nTypeError: Cannot call method 'trigger' of undefined\n\nBacktrace: @file:///home/runner/work/mixxx/mixxx/res/controllers/midi-components-0.0.js:140\ntrigger@file:///home/runner/work/mixxx/mixxx/res/controllers/midi-components-0.0.js:139\nComponent@file:///home/runner/work/mixxx/mixxx/res/controllers/midi-components-0.0.js:53\nButton@file:///home/runner/work/mixxx/mixxx/res/controllers/midi-components-0.0.js:164\n%entry@file:///home/runner/work/mixxx/mixxx/res/controllers/Hercules-P32-scripts.js:109

JoergAtGithub commented:
Finally ready, thanks to the suggestion by @acolombier above!

acolombier commented on res/controllers/midi-components-0.0.js:
```diff
@@ -130,14 +130,18 @@
             if (this.connections[0] !== undefined) {
                 this.connections.forEach(function(conn) {
-                    conn.disconnect();
+                    if (conn !== undefined) {
```
Is this an expected usecase? It does makes sense in the case of testing due to missing CO, but I assume that `connections` should only contain valid connections, else it would indicate a problem. 
Should we add a `console.warn` to flag potentially failed connection?

acolombier commented on src/test/controller_mapping_validation_test.cpp:
```diff
@@ -2,7 +2,10 @@
 #include <gmock/gmock.h>
 #include <gtest/gtest.h>
+#include <qapplication.h>
+#include <qglobal.h>
```
Sorry - I added those imports...
Nit would be to use the alias


```suggestion
#include <QApplication>
#include <Qt>
```

acolombier reviewed (commented):
Looking good, thanks for pushing this to the end! I haven't tested it yet

JoergAtGithub committed:
Remove commented code section

JoergAtGithub committed:
Use Qt include aliases

acolombier reviewed (approved):
LGTM! This seems to slow down the test run by about 30 seconds (for Ubuntu). That being said, the added value is worth the delay IMO. 

Giving some time to let other member a chance to review this as well.
